deeplabcut/pose_estimation_tensorflow/predict_supermodel.py

```diff
@@ -0,0 +1,602 @@
+            if len(ret_pred["coordinates"][0][kpt_id]) == 0:
+                ret_pred["coordinates"][0][kpt_id] = [[]] * len(scale_list)
+                ret_pred["confidence"][kpt_id] = [[]] * len(scale_list)
+
+            temp_coord = np.expand_dims(coord_list, axis=0)
+            ret_pred["coordinates"][0][kpt_id][scale_id] = temp_coord
+            temp_confidence = np.expand_dims(confidence[kpt_id], axis=0)
+            ret_pred["confidence"][kpt_id][scale_id] = temp_confidence
+
+    for kpt_id in range(num_kpts):
+
+        remove_indices = []
+        for idx, ele in enumerate(ret_pred["coordinates"][0][kpt_id]):
+            if len(ele) == 0 or len(ele[0]) == 0:
+                remove_indices.append(idx)
+
+        for idx, ele in enumerate(ret_pred["coordinates"][0][kpt_id]):
+            if idx in remove_indices:
+                # using [0,0] instead of [nan,nan] for cosine similarity to correctly pick up distances
+                ret_pred["coordinates"][0][kpt_id][idx] = np.array([[0, 0]])
+                ret_pred["confidence"][kpt_id][idx] = np.array([[0]])
+
+        mean_vec = np.nanmedian(np.array(ret_pred["coordinates"][0][kpt_id]), axis=0)
+        candidates = np.array(ret_pred["coordinates"][0][kpt_id])
+        dist = []
+
+        for i in range(len(candidates)):
+            # In case where the the predictions do not exist
+            dist.append(cosine_similarity(candidates[i], mean_vec))
+
+        filter_indices = []
+
+        for idx, ele in enumerate(ret_pred["coordinates"][0][kpt_id]):
+            if (
+                dist[idx] < cos_dist_threshold
+                or ret_pred["confidence"][kpt_id][idx] < confidence_threshold
+            ):
+
+                filter_indices.append(idx)
+
+        for idx, ele in enumerate(ret_pred["coordinates"][0][kpt_id]):
+            if idx in filter_indices:
+                ret_pred["coordinates"][0][kpt_id][idx] = np.array([[np.nan, np.nan]])
+                ret_pred["confidence"][kpt_id][idx] = np.array([[np.nan]])
+
+        if len(ret_pred["coordinates"][0][kpt_id]) != 0:
+
+            ret_pred["coordinates"][0][kpt_id] = np.concatenate(
+                ret_pred["coordinates"][0][kpt_id], axis=0
+            )
+            ret_pred["confidence"][kpt_id] = np.concatenate(
+                ret_pred["confidence"][kpt_id], axis=0
+            )
+
+            # need np.array for wrapping the list for evaluation code to work correctly
+            ret_pred["coordinates"][0][kpt_id] = np.array(
+                [np.nanmedian(np.array(ret_pred["coordinates"][0][kpt_id]), axis=0)]
+            )
+            ret_pred["confidence"][kpt_id] = np.array(
+                [np.nanmedian(np.array(ret_pred["confidence"][kpt_id]), axis=0)]
+            )
+        else:
+            ret_pred["coordinates"][0][kpt_id] = np.array([[np.nan, np.nan]])
+            ret_pred["confidence"][kpt_id] = np.array([[np.nan]])
+    return ret_pred
+
+
+def video_inference(
```

Comment:
Should we make this one "private"?

Reply:
Yes we should. I don't know if we want to make it private by putting _ as prefix or simply not import it under deeplabcut module

Reply:
Both I suppose.

Reply:
Usually I don't import those private functions in the __init__.py. We can of course add a _ in the function name